MudInput Dispose JS OnBlur

versile2 opened the pull request:
<!--
MAKE SURE TO READ THE CONTRIBUTING GUIDE BEFORE CREATING A PR
https://github.com/MudBlazor/MudBlazor/blob/dev/CONTRIBUTING.md

Testing sections can be removed for documentation changes
-->

<!-- Provide a general summary of your changes in the Title above -->
<!-- Keep the title short and descriptive, as it will be used as a commit message -->

## Description
<!-- Describe your changes in detail and why. -->
<!-- Note any issues that are resolved by this PR -->
<!-- e.g. resolves #1337 or fixes #9310. -->
PR #10842 does not properly dispose of JsEvent on Re-Render/Dispose. Added method to dispose of the event if needed preventing heap errors since the event is queued up. On a proper onblur the component is already disposed/re-rendered and the manual onblur is stale.
References #10842 
Resolves #11024 
Resolves #11034 

## How Has This Been Tested?
<!-- All PRs should implement unit tests if possible. -->
<!-- Please describe how you tested your changes. -->
<!-- Have you created new tests or updated existing ones? -->
<!-- e.g. unit | visually | none -->
Visually, Console

## Type of Changes
<!-- What type of changes does your code introduce? Put an `x` in only one box that applies best: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (fix or improvement to the website or code docs)

<!-- If you made any visual changes, provide screenshots of before/after. If it has moving parts, please attach a high quality video. -->

## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] The PR is submitted to the correct branch (`dev`).
- [x] My code follows the code style of this project.
- [x] I've added relevant tests.

versile2 committed:
Remove statehaschanged, add dispoal method(s)

versile2 committed:
adjusted unit test to stick to proper button on invokeasync

versile2 committed:
sonarqube

versile2 commented:
ESLint doesn't recognize chaining in javascript, SonarQube requires chaining. hrmm.

ScarletKuro commented on src/MudBlazor/Components/Input/MudInput.razor.cs:
```diff
@@ -373,6 +373,7 @@ or InputType.Time
         /// <inheritdoc />
         protected override async ValueTask DisposeAsyncCore()
         {
+            await JsRuntime.InvokeVoidAsyncWithErrorHandling("mudElementRef.removeOnBlurEvent", ElementReference, _dotNetReferenceLazy);
```
It should be inside `IsJSRuntimeAvailable` otherwise on pre-rendering it will be disposed but the JS runtime is not available there.

versile2 replied:
fixed!

versile2 committed:
move inside IsJSRuntime

versile2 commented:
> Kind of wondering if such un-subscriptions can be managed just by JS side like on `onbeforeunload` and have same effect?

I couldn't find any reference to onbeforeunload that I could follow to see what you are asking, any pointers?

ScarletKuro commented:
> I couldn't find any reference to onbeforeunload that I could follow to see what you are asking, any pointers?

```
addOnBlurEvent(element, dotNetReference) {
    element._mudBlurHandler = function(e) {
        e.preventDefault();
        element.blur();
        if (dotNetReference) {
            dotNetReference.invokeMethodAsync('CallOnBlurredAsync');
        } else {
            console.error("No dotNetReference found for iosKeyboardFocus");
        }
    };

    element.addEventListener('blur', element._mudBlurHandler);

    window.addEventListener('beforeunload', function() {
        element.removeEventListener('blur', element._mudBlurHandler);
    });
}
```

ScarletKuro commented:
Also we need to add `null` check for element https://github.com/MudBlazor/MudBlazor/issues/11024#issuecomment-2731829458

versile2 committed:
add proper null check and undo chaining for eslint

ScarletKuro commented:
> If my understanding is correct onbeforeunload is only going to fire on page change not on re-render. So it won't solve the problem in-page.

I thought that the `Assertion failed - heap is currently locked` happens only on page navigation which is why we unsub on components dispose since usually it's called on page navigation, but well there are other cases like component getting removed from the DOM. Another idea would be `MutationObserver` in JS that detects when elements gets removed in the DOM and then observer it and unsubscribe. But that's just ideas.

versile2 commented:
> > If my understanding is correct onbeforeunload is only going to fire on page change not on re-render. So it won't solve the problem in-page.
> 
> I thought that the `Assertion failed - heap is currently locked` happens only on page navigation which is why we unsub on components dispose since usually it's called on page navigation, but well there are other cases like component getting removed from the DOM. Another idea would be `MutationObserver` in JS that detects when elements gets removed in the DOM and then observer it and unsubscribe. But that's just ideas.

I mean that's all I can recreate, but the other issue is asserting that it fails with regular events causing it to re-render. And since I'm not creating new dispose actions and it follows the lifecycle of the component my preference is to put it here. But if you want it in onbeforeunload it should fix page nav heap errors.

ScarletKuro commented:
> But if you want it in onbeforeunload it should fix page nav heap errors

I actually would want you to try the `MutationObserver` pattern, see MS recommendation:
https://learn.microsoft.com/en-us/aspnet/core/blazor/javascript-interoperability/?view=aspnetcore-9.0#dom-cleanup-tasks-during-component-disposal
> Don't execute JS interop code for DOM cleanup tasks during component disposal. Instead, use the [MutationObserver](https://> developer.mozilla.org/docs/Web/API/MutationObserver) pattern in JavaScript (JS) on the client for the following reasons

Don't need to use `IJSObjectReference` in our case tho.
My long term plan in MudBlazor would be to implement this across the whole codebase: https://github.com/MudBlazor/MudBlazor/pull/11041#issuecomment-2734684070

versile2 commented:
> > But if you want it in onbeforeunload it should fix page nav heap errors
> 
> I actually would want you to try the `MutationObserver` pattern and see how it goes and if it can work in this case. MSFT recommendation and some code samples: https://learn.microsoft.com/en-us/aspnet/core/blazor/javascript-interoperability/?view=aspnetcore-9.0#dom-cleanup-tasks-during-component-disposal
> 
> > Don't execute JS interop code for DOM cleanup tasks during component disposal. Instead, use the [MutationObserver](https://> developer.mozilla.org/docs/Web/API/MutationObserver) pattern in JavaScript (JS) on the client for the following reasons
> 
> Don't need to use `IJSObjectReference` in our case tho. My long term plan in MudBlazor would be to implement this across the whole codebase: [#11041 (comment)](https://github.com/MudBlazor/MudBlazor/pull/11041#issuecomment-2734684070)

After doing the reading and research I believe just a custom element to handle onblur period is the best method using connectedCallback and disconnectedCallback. Since I'm never doing any checking I can remove both onblur "events" and replace with one. It'll be clear in code and javascript I believe. Thoughts?

versile2 committed:
Refactor MudInput for iOS blur handling

Restructure `MudInput.razor` to use a new `<mud-ios-blur-handler>` component for managing special blur events on iOS devices.
Remove outdated JavaScript event handling from `MudInput.razor.cs`, `ElementReferenceExtensions.cs`, and `mudElementReference.js`. Introduce `mudIOSOnBlur.js` to define the `MudIosBlurHandler` class, which ensures proper blur event management while keeping the input visually hidden.

versile2 commented:
I went ahead and did the customElement path with some basic testing. If you approve I'll do full testing which will take quite a bit longer. If not I'll switch to the mutationobserver.

ScarletKuro commented:
```
<mud-ios-blur-handler dotnet-ref="@_dotNetReferenceLazy.Value" assembly="@(typeof(MudInput<T>).Assembly)">
```
Didn't know you can do something like this, where did you find it, I did not see this in MSFT docs.
Very interesting solution, learned something new, but let's revert that last commit, I think the `typeof(MudInput<T>).Assembly` is not very trim friendly thing, especially for something aggressive as iOS AOT.

versile2 commented:
> ```
> <mud-ios-blur-handler dotnet-ref="@_dotNetReferenceLazy.Value" assembly="@(typeof(MudInput<T>).Assembly)">
> ```
> 
> Didn't know you can do something like this, where did you find it, I did not see this in MSFT docs. Very interesting solution, learned something new, but let's revert that last commit, I think the `typeof(MudInput<T>).Assembly` is not a very trim friendly thing, especially for something aggressive as iOS AOT.

Made a proper example because.. Well I was curious. But I just read MSFT link you supplied... hehe
https://github.com/versile2/Custom-HTML-Element

versile2 committed:
Revert "Refactor MudInput for iOS blur handling"

This reverts commit 3dffbcd924c449ee9577e5f057e32bd63ea4db6f.

Cobra86 commented:
I'm still getting the error after 8.5  Issue #11034
I did caching thing and still same.
`<script src="_content/MudBlazor/MudBlazor.min.js?v=8.5"></script>`

versile2 commented:
> I'm still getting the error after 8.5 Issue #11034 I did caching thing and still same. `<script src="_content/MudBlazor/MudBlazor.min.js?v=8.5"></script>`

Give us something we can reproduce and I'll look at it asap. Add it to the issue, it'll notify me. If you can't do it in trymud then make a github project.

versile2 commented:
That's for the screenshot, I was able to see you were using Edge. I was able to use your try snippet in edge on version 8.4 and reproduce the error. However, using version 8.51 the error is gone.
https://try.arctechonline.tech/snippet/GaGTaIYvqppqpGjC
